elastiknn-api4s/src/main/scala/com/klibisz/elastiknn/api/XContentCodec.scala

```diff
@@ -0,0 +1,598 @@
+          case q: NearestNeighborsQuery.JaccardLsh     => jaccardLshQuery.encodeUnsafe(q, b)
+          case q: NearestNeighborsQuery.HammingLsh     => hammingLshQuery.encodeUnsafe(q, b)
+          case q: NearestNeighborsQuery.CosineLsh      => cosineLshQuery.encodeUnsafe(q, b)
+          case q: NearestNeighborsQuery.L2Lsh          => l2LshQuery.encodeUnsafe(q, b)
+          case q: NearestNeighborsQuery.PermutationLsh => permutationLshQuery.encodeUnsafe(q, b)
+        }
+    }
+  }
+
+  object Decoder {
+
+    private def unexpectedName(current: String): String =
+      s"Unexpected name [$current]"
+
+    private def unexpectedValue(s: String): String =
+      s"Unexpected value [$s]"
+
+    private def unexpectedToken(current: Token, expected: Token*): String =
+      s"Expected token to be one of [${expected.mkString(",")}] but found [${current}]"
+
+    private def unableToConstruct(tipe: String): String =
+      s"Unable to construct [$tipe] from parsed JSON"
+
+    private def assertToken(current: Token, expected: Token*): Unit =
+      if (expected.contains(current)) () else throw new XContentParseException(unexpectedToken(current, expected: _*))
+
+    private def parseFloatArray(p: XContentParser, expectedLength: Int): Array[Float] = {
+      val b = new ArrayBuffer[Float](expectedLength)
+      p.currentToken() match {
+        case Token.START_ARRAY  => ()
+        case Token.VALUE_NUMBER => b.append(p.floatValue())
+        case t                  => throw new XContentParseException(unexpectedToken(t, Token.START_ARRAY, Token.VALUE_NUMBER))
+      }
+      while (p.nextToken() != Token.END_ARRAY) b.append(p.floatValue())
+      b.toArray
+    }
+
+    private def parseSparseBoolArray(p: XContentParser, expectedLength: Int): Array[Int] = {
+      val b = new ArrayBuffer[Int](expectedLength)
+      p.currentToken() match {
+        case Token.START_ARRAY  => ()
+        case Token.VALUE_NUMBER => b.append(p.intValue())
+        case t                  => throw new XContentParseException(unexpectedToken(t, Token.START_ARRAY, Token.VALUE_NUMBER))
+      }
+      while (p.nextToken() != Token.END_ARRAY) b.append(p.intValue())
+      b.toArray
+    }
+
+    implicit val similarity: Decoder[Similarity] = (p: XContentParser) => {
+      if (p.currentToken() != Token.VALUE_STRING) assertToken(p.nextToken(), Token.VALUE_STRING)
+      val s1 = p.text()
+      val s2 = s1.toLowerCase
+      s2 match {
+        case Names.JACCARD => Similarity.Jaccard
+        case Names.HAMMING => Similarity.Hamming
+        case Names.L1      => Similarity.L1
+        case Names.L2      => Similarity.L2
+        case Names.COSINE  => Similarity.Cosine
+        case Names.ANGULAR => Similarity.Cosine
+        case _             => throw new XContentParseException(unexpectedValue(s1))
+      }
+    }
+
+    implicit val vec: Decoder[Vec] = new Decoder[Vec] {
+      override def decodeUnsafe(p: XContentParser): Vec = {
+        var field: Option[String] = None
```

Comment:
The recipe for decoders with the XContentParser seems to be:

1. Define a var of an Option for every member of the object that could be populated.
2. Step through the tokens, setting the Options for each member encountered.
3. Pattern match on the Options to determine and construct the corresponding scala case class.